differs/emerge_diff.go

```diff
@@ -0,0 +1,117 @@
+/*
+Copyright 2018 Google, Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package differs
+
+import (
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	pkgutil "github.com/GoogleContainerTools/container-diff/pkg/util"
+	"github.com/GoogleContainerTools/container-diff/util"
+	"github.com/sirupsen/logrus"
+)
+
+//Emerge package database location
+const emergePkgFile string = "/var/db/pkg"
+
+type EmergeAnalyzer struct {
+}
+
+func (em EmergeAnalyzer) Name() string {
+	return "EmergeAnalyzer"
+}
+
+// EmergeDiff compares the packages installed by apt-get.
+func (em EmergeAnalyzer) Diff(image1, image2 pkgutil.Image) (util.Result, error) {
+	diff, err := singleVersionDiff(image1, image2, em)
+	return diff, err
+}
+
+func (em EmergeAnalyzer) Analyze(image pkgutil.Image) (util.Result, error) {
+	analysis, err := singleVersionAnalysis(image, em)
+	return analysis, err
+}
+
+func (a EmergeAnalyzer) getPackages(image pkgutil.Image) (map[string]util.PackageInfo, error) {
+	var path string
+	path = image.FSPath
+	switch path {
+	case "":
+		path = emergePkgFile
+	default:
+		path = filepath.Join(path, emergePkgFile)
+	}
+	packages := make(map[string]util.PackageInfo)
+	if _, err := os.Stat(path); err != nil {
+		// invalid image directory path
+		logrus.Errorf("Invalid image directory path %s", path)
+		return packages, err
+	}
+
+	contents, err := ioutil.ReadDir(path)
+	if err != nil {
+		logrus.Errorf("Non-content in image directory path %s", path)
+		return packages, err
+	}
+
+	for i := 0; i < len(contents); i++ {
+		c := contents[i]
+		pkgPrefix := c.Name()
+		pkgContents, err := ioutil.ReadDir(filepath.Join(path, pkgPrefix))
+		if err != nil {
+			return packages, err
+		}
+		for j := 0; j < len(pkgContents); j++ {
+			c := pkgContents[j]
+			pkgRawName := c.Name()
+			s := strings.Split(pkgRawName, "-")
+			if len(s) != 2 {
+				continue
+			}
+			pkgName, version := s[0], s[1]
+			pkgPath := filepath.Join(path, pkgPrefix, pkgRawName, "SIZE")
+			size := getPkgSize(pkgPath)
+			currPackage := util.PackageInfo{Version: version, Size: size}
+			fullPackageName := strings.Join([]string{pkgPrefix, pkgName}, "/")
+			packages[fullPackageName] = currPackage
+		}
+	}
+
+	return packages, nil
+}
+
+func getPkgSize(pkgPath string) int64 {
+	var sizeFile *os.File
```

Comment:
Perhaps I am missing something, but can this just use os.Stat to get the file size? https://gist.github.com/miguelmota/08bca5febf2ce460a289ef14c7e9af4f

If not, maybe a comment that says

`//  getPkgSize does X because it's special`

Reply:
Well, this SIZE file and created by emerge, since emerge will count the total size of a package and store it as a SIZE file in packages metadata directory. 

And it's quite complex using `os.Stat` to get the file size of a package, because all dependency files are scattered. For example, if installing `dev-python/mysqlclient-3.0.0`, it will generate a link file `/usr/lib/_mysql.so`, client codes in `/usr/lib64/python3.7/site-packages/mysqlclient/` and python egg file in `/usr/lib64/python3.7/` . Obviously, it's not easy to get size of all these files.